Source/Shared/arcana/threading/cancellation.h

```diff
@@ -28,22 +28,17 @@ namespace arcana
             template<typename CallableT>
             ticket add_listener(CallableT&& callback, std::function<void()>& copied)
             {
-                std::scoped_lock<std::mutex> lock{ m_mutex };
+                return internal_add_listener(m_cancelStartedListeners, std::forward<CallableT>(callback), copied);
+            }
 
-                if (m_cancelFinished)
-                {
-                    copied = std::forward<CallableT>(callback);
-                    return m_listeners.insert(copied, m_mutex);
-                }
-                else
-                {
-                    return m_listeners.insert(std::forward<CallableT>(callback), m_mutex);
-                }
+            template<typename CallableT>
+            ticket add_cancellation_completed_listener(CallableT&& callback, std::function<void()>& copied)
```

Comment:
I think maybe just this without `cancellation` since this is the cancellation class already.
```suggestion
            ticket add_completed_listener(CallableT&& callback, std::function<void()>& copied)
```